### PR TITLE
Move WMR behind a dedicated feature flag

### PR DESCRIPTION
**Pull request by aalexandrov**
This introduces a dedicated feature flag for the WMR syntax. The `--unsafe-mode` toggle continues to work as before.

### Motivation

  * This PR adds a known-desirable feature.

Design doc PR: MaterializeInc/materialize#17820.
Closes MaterializeInc/database-issues#5226.

### Tips for reviewer

- The first commit adds a new feature `enable_with_mutually_recursive` as well as a new `SessionCatalog::get_feature` method that allows for type-safe lookup of a selected subset of the `SystemVars` fields of type `ServerVar<bool>`.
- The second commit adds a complementary `SessionCatalog::set_feature` method which is called in `CatalogState::parse_view_item` and `Catalog::parse_item` in order to ensure that existing catalog items that rely on a SQL-level feature are still supported even if the feature is turned on.

We need to decide whether we want to include both commits or only the first one.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/adapter/src/catalog.rs
Comment (aalexandrov) [inline src/adapter/src/catalog.rs:6669]: At the moment the `soft_assert` ensures that the `feature_name` value references an actual `SystemVar<bool>` field.
I can also change the signature of the enclosing method if there is a way to convert an `AdapterError` into a `SqlCatalogError`.

File: src/adapter/src/catalog.rs
Comment (aalexandrov) [inline src/adapter/src/catalog.rs:6670]: Using `false` as the default seems like the safer option. The `soft_assert` above will hopefully ensure that we catch any dangling `is_enabled` calls to no longer existing feature flags in Sentry.

File: src/sql/src/plan/query.rs
Comment (aalexandrov) [inline src/sql/src/plan/query.rs:1178]: I will introduce a constant. Ideally this should be the `static WITH_MUTUALLY_RECURSIVE: ServerVar<bool>` from `vars.rs`, but the `adapter` code is not available in `mz_sql`. I think in the long term we should try to move `vars.rs` to some crate that is further up the dependency graph.

File: src/sql/src/plan/query.rs
Comment (aalexandrov) [inline src/sql/src/plan/query.rs:1178]: Note that I have introduce a dedicated constant for the feature flag name under `src/sql/lib.rs`.

Ideally this should be referencing the `name()` of `static WITH_MUTUALLY_RECURSIVE: ServerVar<bool>` from `vars.rs`, but the `adapter` code is not available in `mz_sql`. I think in the long term we should try to move `vars.rs` to some crate that is further up the dependency graph.

File: src/sql/src/lib.rs
Comment (jkosh44) [reply src/sql/src/lib.rs:178]: Oh, I just saw your other comment. Does `vars.rs` have access to this crate so at least we can use this constant in the variable itself? I think this is very fragile, if someone changes the name in `vars.rs` and forgets to change this then `is_enabled` with this always returns false.

File: src/adapter/src/session/vars.rs
Comment (antiguru) [reply src/adapter/src/session/vars.rs:415]: I wonder if it should be `enable_with_mutually_recursive`, which would indicate what the flag does?

Comment (antiguru) [reply src/adapter/src/session/vars.rs:415]: I wonder if it should be `enable_with_mutually_recursive`, which would indicate what the flag does?

File: src/sql/src/lib.rs
Comment (aalexandrov) [reply src/sql/src/lib.rs:178]: `vars.rs` is not available in the `sql/plan/statement` which is the sole user of this at the moment.

We won't need this if we move forward with the `is_{feature}_enabled` methods. Maybe I should rework the PR until the pain of maintaining one `StatementContext` method per feature flag is bit enough so we can invest in moving `vars.rs` to a better place.

File: src/adapter/src/session/vars.rs
Comment (aalexandrov) [reply src/adapter/src/session/vars.rs:415]: OK, then we can name feature flags like `enable_{feature}` and have a corresponding method on `StatementContext` called `require_{feature}` that statically checks the value of `enable_{feature}` in the backing `self.catalog`.

If we all agree I will do it like that and I'll update the notion doc so this is done consistently for all backend features.

Comment (aalexandrov) [reply src/adapter/src/session/vars.rs:415]: cc @ggnall regarding :point_up:.

File: src/sql/src/lib.rs
Comment (jkosh44) [reply src/sql/src/lib.rs:178]: > We won't need this if we move forward with the is_{feature}_enabled methods. Maybe I should rework the PR until the pain of maintaining one StatementContext method per feature flag is bit enough so we can invest in moving vars.rs to a better place.

I think that's a good idea. I can make an issue for the moving of `vars.rs`. That'll be a good first issue too for some of the new folks.

File: src/adapter/src/session/vars.rs
Comment (ggnall) [reply src/adapter/src/session/vars.rs]: 👍 on the flag naming convention

File: src/adapter/src/session/vars.rs
Comment (philip-stoev) [inline src/adapter/src/session/vars.rs:413]: copypasta in comment

**Comment by aalexandrov** [comment]
@jkosh44 I reworked the code to be a bit more type-safe. The downside is that `catalog.rs` now contains an enum (`EnableWithMutuallyRecursive`) that replicates some of the constants that live in `vars.rs`. This can probably be removed if we make them available to the `sql` crate in the future.

File: src/adapter/src/catalog.rs
Comment (aalexandrov) [reply src/adapter/src/catalog.rs]: I went with a

```rust
/// Returns `true` iff the given `feature` is supported at the moment.
fn supports_feature(&self, feature: CatalogFeature) -> bool;
```

method where `CatalogFeature` is defined as follows:

```rust
/// Defines features that the [SessionCatalog] may optionally support.
#[derive(Debug)]
pub enum CatalogFeature {
    /// True iff `WITH MUTUALLY RECURSIVE` syntax is enabled.
    EnableWithMutuallyRecursive,
}
```

File: src/sql/src/lib.rs
Comment (aalexandrov) [reply src/sql/src/lib.rs]: Done.

**Comment by aalexandrov** [comment]
@antiguru note that at the moment the feature handling mechanism breaks if the feature is not enabled but the catalog contains objects that contain WMR syntax. I can probably invest more time to do this as well, but it will probably require at least 1-2 more days. LMK if you feel this is important.

File: src/adapter/src/catalog.rs
Comment (parker-timmerman) [inline src/adapter/src/catalog.rs:6155]: nit: it would be more idiomatic Rust to add a `with_feature(self, feature: CatalogFeature)` method on `ConnCatalog`, which does what this is doing, enables the feature if it's not already enabled, this way you could use the builder pattern like:

```
let session_catalog = self.for_system_session()
                          .with_feature(CatalogFeature::EnableWithMutuallyRecursive);
```

Comment (aalexandrov) [reply src/adapter/src/catalog.rs:6155]: `enable_features_required_for_catalog_open` needs to be called twice, and over time I suspect the number of features that evolve will change (some will be removed and others will be added depending on the current epics and the lifecycle of each feature).

I therefore prefer to have a single method which lives in `catalog.rs` which is where the decision of which of the features need to be enabled for item parsing needs to be made.

I can define it like that:

```rust
fn enable_features_required_for_catalog_open(session_catalog: ConnCatalog) -> ConnCatalog {
    session_catalog
        // enable features required for parsing
        .with_feature(CatalogFeature::EnableWithMutuallyRecursive)
}
```

File: src/adapter/src/session/vars.rs
Comment (jkosh44) [inline src/adapter/src/session/vars.rs:1590]: Does `VarInput::Flat(value.format())` work instead of `VarInput::Flat(if value { "true" } else { "false" })` since `Value` is implemented for `bool` below?

File: src/adapter/src/catalog.rs
Comment (jkosh44) [inline src/adapter/src/catalog.rs:5821]: Doesn't this mean that the feature gets turned on for every environment, every time it starts up?

Comment (aalexandrov) [reply src/adapter/src/catalog.rs:5821]: Yes, but AFAICT the `session_catalog` constructed here is only used in the
- `mz_sql::parse::parse`,
- `mz_sql::names::resolve`, and 
- `mz_sql::plan::plan`
calls of the enclosing function (`parse_item` / `parse_view_item`) and then destroyed, so this override should only affect the scope of `Catalog::open`. For regular client connections the original value should be still be used.

File: src/adapter/src/session/vars.rs
Comment (aalexandrov) [reply src/adapter/src/session/vars.rs:1590]: Ah, nice catch!

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:5821]: But the system variables are scoped to the entire catalog, not just an individual session. So doesn't this change the value for all future sessions too?

File: src/adapter/src/catalog.rs
Comment (jkosh44) [inline src/adapter/src/catalog.rs:6709]: As per the other conversation, this modifies the catalog state, which is used by all current and future sessions.

Comment (aalexandrov) [reply src/adapter/src/catalog.rs:6709]: I thought that `self.state.to_mut()` is copy-on-write, so it will just clone the state for this `ConnCatalog` instance. Let me test this.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:5821]: Perhaps we should only enable it if it's disabled and then after re-disable it.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:6709]: > I thought that `self.state` is copy-on-write (`Cow<_>`), so `self.state.to_mut()` will just clone the state for this `ConnCatalog` instance. Let me test this.

Oh, you're absolutely right. I didn't realize that it's a `Cow`. This should be fine then.

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/adapter/src/catalog.rs
Comment (aalexandrov) [reply src/adapter/src/catalog.rs:6709]: I tested and after I restart `bin/environmentd` a catalog entry that uses WMR is re-created, but 
```sql
show enable_with_mutually_recursive;
```
still returns `false`. If @antiguru wants this behavior I will add an `mzcompose` test for this scenario.

File: src/adapter/src/catalog.rs
Comment (aalexandrov) [reply src/adapter/src/catalog.rs:5821]: > Perhaps we should only enable it if it's disabled

I am only doing this in order to avoid cloning if not needed (see the guards in `enable_features_required_for_catalog_open`). It will still affect most `environmentd` instances in production, though...

> and then after re-disable it.

I'm not sure if this is needed if we are going to drop the clone either way at the end of the method. It's more future proof but then I need two methods instead of one and I need to around the set of features that need to be flipped as state.

File: src/adapter/src/catalog.rs
Comment (aalexandrov) [reply src/adapter/src/catalog.rs:6709]: > Also this is the only place where we modify a SystemVar in memory, without also modifying it in the stash. That's probably fine but we should call this out in some doc-comment with an explanation as to why this is.

I'll extend the `set_feature` docs so people are not tricked of assuming that these calls are being persisted.

**Comment by antiguru** [comment]
> @antiguru note that at the moment the feature handling mechanism breaks if the feature is not enabled but the catalog contains objects that contain WMR syntax. I can probably invest more time to do this as well, but it will probably require at least 1-2 more days. LMK if you feel this is important.
> 
> _update_: This was a bit easier than expected. The second commit now does this. We should decide whether we want this behavior or not. This was the case for `window_functions` feature flag #15969, but there we had customers already using the feature in production when we rolled it out.
> 
> Once we agree on this behavior, I will also add a specific test (if we want the feature to be interpreted in a `Catalog::open`-safe manner I'll add an `mzcompose.py`, otherwise an `*.slt` is sufficient).

Me trying to shine some light on this:
* For internal users, we know that it's an unstable feature and we can expect to let people know what that entails.
* For external users, we have to make it very clear that we can disable the feature at any point, and we have to communicate what this means. Do we just not install the catalog item? Do we wipe the environment? I don't know!
* If it's easy to not break the catalog when turning the feature off, then let's do this.

**Comment by aalexandrov** [comment]
Summoning @ggnall for his opinion.

Basically what we want to prevent are situations where `environmentd` fails to start because somebody disabled a feature flag that affects if `WITH MUTUALLY RECURSIVE` is enabled or not in the presence of views / materialized views that already use the syntax. In theory this should not happen (that is, customers should be notified and should clear their catalog state of items that use a specific feature before we turn it off).

> Me trying to shine some light on this:
> 
> * For internal users, we know that it's an unstable feature and we can expect to let people know what that entails.
> * For external users, we have to make it very clear that we can disable the feature at any point, and we have to communicate what this means. Do we just not install the catalog item? Do we wipe the environment? I don't know!
> * If it's easy to not break the catalog when turning the feature off, then let's do this.

**Comment by aalexandrov** [comment]
@ggnall another possible option to handle a situation where we have a `WMR view / materialized view in the catalog but the feature is not enabled is to decide not to install it and raise an internal error. However, this can still lead to invalid catalog state if there are downstream items that depend on the offending (WMR-using) item.

**Comment by aalexandrov** [comment]
> If it's easy to not break the catalog when turning the feature off, then let's do this.

I don't think this is easy because we will need to reason about dependencies between catalog items while re-hydrating the catalog .

File: ci/nightly/pipeline.yml
Comment (philip-stoev) [inline ci/nightly/pipeline.yml:62]: Can you please name the directory and the buildkite step it "sql-feature-flags" or even just `feature-flags`. This way there will be no confusion as to what exactly is being tested here.

File: misc/python/materialize/mzcompose/services.py
Comment (aalexandrov) [inline misc/python/materialize/mzcompose/services.py:81]: @philip-stoev this is needed because the `sql-features/mzcompose.py` needs to start `Materialized` without this flag (otherwise all SQL-level feature flags will be always `on`).

File: test/sql-features/mzcompose.py
Comment (philip-stoev) [inline test/sql-features/mzcompose.py:68]: Generate

File: test/sql-features/mzcompose.py
Comment (aalexandrov) [inline test/sql-features/mzcompose.py:95]: The essence of the test logic is in the `phase1` and `phase2` methods and the 
```python
for scenario in scenarios:
   # ...
```
loop in `run_test` below.

File: test/sql-features/mzcompose.py
Comment (philip-stoev) [inline test/sql-features/mzcompose.py]: There is no longer a need to create temporary files, you can simply do:

```
c.testdrive(input="your string goes here")
```

My suggestion is to simply pass `c: Composition` to all code that is currently generating testdrive fragments. That code can then use `c.testdrive` to simply run the fragment right away without having to pass it around.

File: test/sql-features/mzcompose.py
Comment (philip-stoev) [inline test/sql-features/mzcompose.py:69]: It took me a while to figure out what is going on here and all the other places where a similar construct is used -- you are fishing for the second line returned by the `create_item()` method. It may be best to split this into two methods -- `create_item()` will return the `CREATE` portion and then an `invoke()` method can return the `SELECT` portion , which you can then use directly without fishing for it.

File: test/sql-features/mzcompose.py
Comment (philip-stoev) [inline test/sql-features/mzcompose.py:92]: Optionally you may save some boilerplate by using https://docs.python.org/3/library/dataclasses.html

File: test/sql-features/mzcompose.py
Comment (philip-stoev) [inline test/sql-features/mzcompose.py:40]: Technically speaking you do not need to do that -- just never drop the schema throughout the test.

**Review by philip-stoev — APPROVED** [review]

File: test/sql-features/mzcompose.py
Comment (aalexandrov) [reply test/sql-features/mzcompose.py]: The fragments are quite longa and enclosing them in tempfile creates some nice sections in the build:

https://buildkite.com/materialize/nightlies/builds/1912#0186c12d-1188-45f0-bfee-bc913c96bcd9

File: test/sql-features/mzcompose.py
Comment (aalexandrov) [reply test/sql-features/mzcompose.py:69]: I need to transform a possibly mutli-line SQL statement such as

```sql
CREATE TABLE t (
  x INT,
  y INT
)
```
to a valid TD command
```
> CREATE TABLE t (
    x INT,
    y INT
  )
```
which boils down to (1) indent the first line with '> ', and (2) indent all other lines with `  `. I'll see if this can be more readable.

File: test/sql-features/mzcompose.py
Comment (aalexandrov) [reply test/sql-features/mzcompose.py:40]: If we have multiple test scenarios it might be better to start each scenario on a clean slate.

File: test/sql-features/mzcompose.py
Comment (philip-stoev) [reply test/sql-features/mzcompose.py]: I think the caller needs to supply the statement fully-formatted so that you do not need to do that surgery.

Alternatively, you can run the statement using `$ postgres-execute` in which case everything up to the next `$` directive will be slurped and executed together.

Comment (aalexandrov) [reply test/sql-features/mzcompose.py]: > I think the caller needs to supply the statement fully-formatted so that you do not need to do that surgery.

This violates the implicit contract of these helper functions: they receive data (SQL statements, expected errors, key-value pairs) and construct specific testdrive commands.

> Alternatively, you can run the statement using `$ postgres-execute` in which case everything up to the next $ directive will be slurped and executed together.

This is fine for the `statement_ok` function but not for the `statement_error` function. How about:

```python
indent(statement.strip(), prefix="  ").replace("  ", "> ", 1)
```
and
```python
indent(statement.strip(), prefix="  ").replace("  ", "! ", 1)
```

Comment (philip-stoev) [reply test/sql-features/mzcompose.py]: Thanks, understood.

**Comment by ggnall** [comment]
> In theory, I think that this should not happen (that is, customers should be notified and should clear their catalog state of items that use a specific feature before we turn it off). However, it seems better be safe here, even at the risk of handling some catalog items that should not be handled based on the current feature flag configuration.

Agree on taking the safe path here With all experimental features, we do reserve the ability to tell users they must delete objects and then turn the feature off for them. But we should be defensive about cases where the flag setting could crash loop an environment. 

If we have some pressure on release timing, I think it's acceptable to not solve this ahead of internal users having access to the flag, but it should definitely be protected against before any external user has access.

**Comment by aalexandrov** [comment]
> If we have some pressure on release timing, I think it's acceptable to not solve this ahead of internal users having access to the flag, but it should definitely be protected against before any external user has access.

@ggnall I already did this, just waiting for an ack from you before I merge the PR with this behavior implemented.

**Review by ggnall — APPROVED** [review]

**Pull request merged**
